src/main/utility/once_set.rs

```diff
@@ -0,0 +1,61 @@
+use std::collections::HashSet;
+use std::sync::RwLock;
+
+/// A [`HashSet`] that only allows insertions and uses interior mutablity. This allows it to be used
+/// in a global static.
+#[derive(Debug, Default)]
+pub struct OnceSet<T>(RwLock<Option<HashSet<T>>>);
+
+impl<T> OnceSet<T>
+where
+    T: std::cmp::Eq + std::hash::Hash,
+{
+    pub const fn new() -> Self {
+        Self(RwLock::new(None))
+    }
+
+    /// Insert `val` into the set. Returns `false` if `val` had previously been added to the set;
+    /// otherwise returns `true`.
+    pub fn insert(&self, val: T) -> bool {
+        // First check with a (cheap) read-lock.
+        let already_added = self
```

Comment:
Optional nit - could get rid of the already_added` variable, simplifying this function a little bit.

```
  if self./*...*/.unwrap_or(false) {
    return false;
  }
  
  self.0./*...*/.insert(val)
}
```

Reply:
Good idea, added!